### ISSUE TITLE
Support DuckDB

### ISSUE THREAD

**Issue by chriszrc** [prisma/prisma#21281]
## Problem
DuckDB is a superior solution for an embedded db solution, especially compared to sqlite:

https://duckdb.org/docs/api/nodejs/overview.html

Currently if packaging a version of my application for desktop/offline/electron use, I'd need to use sqlite to continue using prisma, but I am trying to avoid using sqlite if possible-

## Suggested solution
Add support for duckdb

## Alternatives
n/a

## Additional context
n/a

**Comment by janpio** [comment]
Is there a definition of the SQL (dialect) DuckDB uses?

I found [this comment from 2020](https://news.ycombinator.com/item?id=24531894) which says `we use the PostgreSQL parser, and try to be generally compatible with the syntax used by SQLite, Postgres and MySQL`, but nothing current in the docs. If we can use one of our existing dialects (PostgreSQL, MySQL or SQLite) to directly talk to the DuckDB driver, this will be _very_ possible.

**Comment by chriszrc** [comment]
@janpio Hi, yes, they really do use the postgres parser:

https://duckdb.org/why_duckdb#:~:text=and%20Panos%20Kalnis.-,SQL%20Parser%3A,-We%20use%20the

And my experience has been that sticking with the postgres syntax has been seamless. Of course duckdb includes things that postgres hasn't implemented, like `select * exclude()`, but even just being able to have an embedded db that has parity with postgres would be amazing. 

Schema migrations work just like postgres, unless maybe I'm misunderstanding your question, is there something in particular you're wondering about?

**Comment by janpio** [comment]
Then it sounds like DuckDB support could maybe be achieved via our upcoming "driver adapters" feature. Then Prisma Client can just use the Node.js API you linked to above to run queries. Take a look at https://www.prisma.io/blog/prisma-turso-ea-support-rXGd_Tmy3UXX for a first example that we just released yesterday. Migrations and Introspection is still undefined how those Driver Adapters would work in Prisma CLI, so we would probably start with a workaround similar to the Turso one: Use `migrate diff` or `dev --create-only` to create the SQL, but then execute it some other way. But we'll figure that out as well soon.

**Comment by reza-ebrahimi** [comment]
For reference regarding the Prisma priority: https://github.com/prisma/prisma-engines/issues/3106#issuecomment-1209335602

>Unfortunately, our Product team has set a very clear direction that we're not going to invest any time, neither into building new DB connectors, nor Prisma Client for new languages in the next couple of months. I still think it's great to have this issue open as a future reminder for the time when we're ready to expand into other languages, but for now I just want to set expectations from our side that this is not going to end up on our roadmap any time soon, hope you understand!

**Comment by janpio** [comment]
That statement is over 1 year old @reza-ebrahimi, as you can see with recent launches like Turso at https://www.prisma.io/blog/prisma-turso-ea-support-rXGd_Tmy3UXX, we are adding support for new databases - just not the way  it was requested in the issue you link to.

**Comment by kuatroka** [comment]
+1 for duckdb

**Comment by kavenugop** [comment]
@janpio - How is the support for duckdb expected based on that article? Assuming its via driverAdapters?  Can you please elaborate.

**Comment by janpio** [comment]
Could you rephrase your question? I am not sure what you are asking.

Generally: Database that speak one of the SQL dialects we already support (PostgreSQL, MySQL, SQLite or SQL Server) can get a driver adapter via their Node.js driver / Client for Prisma Client. Only database that use different types and SQL dialect only can work with a full implementation on our side, which is a _lot_ more effort.

**Comment by kavenugop** [comment]
@janpio - It makes sense to use PostgreSQL dialect to connect to DuckDB and I am good if that works. However I have been searching for the node js driver/url to connect to DuckDB and have not had success so far. 

generator client {
  provider             = "prisma-client-py"
}
datasource db {
  provider = "postgres"
  url      = "??duckdb_url??"
}

**Comment by janpio** [comment]
It does not exist right now, someone has to build it first. Here is the documentation of the currently existing driver adapters: https://www.prisma.io/docs/concepts/components/database-drivers#driver-adapters

**Comment by chriszrc** [comment]
@janpio Can you point to where the turso driver/client is located, so we can look at what's involved to get this working for duckdb?

**Comment by janpio** [comment]
https://github.com/search?q=repo%3Aprisma%2Fprisma%20turso&type=code or https://github.com/search?q=repo%3Aprisma%2Fprisma%20libsql&type=code

**Comment by tobilg** [comment]
Is there some info about the adapter API that would need to be implemented @janpio? I can read the code, yes, but it'd be a great enable for potential community contributions to have more info about the API. Thanks!

**Comment by janpio** [comment]
No, the API is shown in the examples we have for now.

**Comment by shariquealavi** [comment]
Do we have duckdb support?

**Comment by ZhangPeng4242** [comment]
DuckDB please!

**Comment by szawadski** [comment]
Also looking forward to getting support fir duckdb

**Comment by jkomyno** [comment]
Hi there,

As part of our effort to clean up our GitHub repo, we’re closing this issue to help focus on features and improvements that align with our focus areas and those of our community. 

If this is issue is still important to you, feel free to reopen the issue and get like-minded users to upvote as we will prioritize fixing issues based on community interest. If you do reopen the issue, please ensure that there’s a clear method for us to reproduce it. 

For more insight into our priorities and how we’re thinking about the future of the Prisma ORM, please see our recent blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and support!
